File: crates_usage/examples/try_login.py
#!/usr/bin/env python

import requests
import base64

from dotenv import load_dotenv
from os import environ as env
from pprint import pprint

load_dotenv() # take environment variables from .env

GOTIFY=f"https://{env.get('BASE_URL')}"

def create_client_basic_auth(username: str, password: str):
    session = requests.Session()
    session.auth = (username, password)
    
    return session.post(
        url=f"{GOTIFY}/client",
        json={
            "name": "Gotify Rustop"
        },
        verify=False, ### INSECURE ###
    ).json()
    
def create_client_base64(username: str, password: str):
    auth = base64.b64encode(f"{username}:{password}".encode("utf-8"))
    auth = auth.decode("utf-8")
    print(auth)
    
    return requests.post(
        url=f"{GOTIFY}/client",
        headers={
            "Authorization": f"Basic {auth}",
        },
        json={
            "id": None,
            "name": "Gotify Rustop",
            "token": None,
        },
        verify=False, ### INSECURE ###
    ).json()
    
def get_current_user(token: str):
    return requests.get(
        url=f"{GOTIFY}/current/user",
        headers={
            "X-Gotify-Key": token,
        },
        verify=False, ### INSECURE ###
    ).json()
    
def delete_client(token: str, id: int):
    return requests.delete(
        url=f"{GOTIFY}/client/{id}",
        headers={
            "X-Gotify-Key": token,
        },
        verify=False, ### INSECURE ###
    )
    
if __name__ == "__main__":
    print("-- Create client --")
    client: list = create_client_basic_auth(input("Username: "), input("Password: "))
    pprint(client)
    
    token = client["token"]
    
    print("-- Get current user --")
    current_user = get_current_user(token)
    pprint(current_user)
    
    print("-- Delete client --")
    response = delete_client(token, client["id"])
    pprint(response)

File: crates_usage/examples/try_login.rs
use std::env;
use std::io::stdin;
use dotenv::dotenv;

use reqwest::Response;
use serde_derive::{Serialize, Deserialize};
use serde_json::Value;
use serde_json::value::from_value;

use lazy_static::lazy_static;

lazy_static! {
    static ref GOTIFY: String = format!("https://{}", env::var("BASE_URL").unwrap());
}
//
#[tokio::main]
async fn main() {
    dotenv().ok(); // Load env variables in .env
    
    println!("-- Create client --");
    let mut username = String::new();
    println!("Username: ");
    stdin().read_line(&mut username).unwrap();

    let mut password = String::new();
    println!("Password: ");
    stdin().read_line(&mut password).unwrap();

    let client = create_client(
            username.as_str().trim(),
            password.as_str().trim(),
        ).await.unwrap();
    println!("{:#?}", &client);

    let token = &client["token"];
    println!("{}", &token);


    println!("\n-- Get current user --");
    let current_user = get_current_user(&token.to_string()).await.unwrap();
    println!("{:#?}", &current_user);

    
    println!("\n-- Delete client --");
    let response = delete_client(
            &token.to_string(),
            from_value::<i32>(client["id"].clone()).unwrap(),
        ).await.unwrap();
    println!("{:#?}", &response);
}

#[derive(Serialize, Deserialize)]
struct ClientModel {
    id: Option<i32>,
    name: String,
    token: Option<String>,
}

impl ClientModel {
    fn new(name: &str) -> Self {
        Self {
            id: None,
            name: name.to_string(),
            token: None,
        }
    }
}

async fn create_client(username: &str, password: &str) -> Result<Value, Box<dyn std::error::Error>> {   
    let body = ClientModel::new("Gotify Rustop");

    let client = reqwest::Client::new()
        .post(format!("{}/client", *GOTIFY))
        .basic_auth(username, Some(password))
        .json::<ClientModel>(&body);

    let resp = client.send()
        .await?
        .json::<Value>()
        .await?;

    Ok(resp)
}

async fn get_current_user(token: &str) -> Result<Value, Box<dyn std::error::Error>> {
    let client = reqwest::Client::new()
        .get(format!("{}/current/user", *GOTIFY))
        .header("X-Gotify-Key", token);

    let resp = client.send()
        .await?
        .json::<Value>()
        .await?;

    Ok(resp)
}

async fn delete_client(token: &str, id: i32) -> Result<Response, Box<dyn std::error::Error>> {
    let client = reqwest::Client::new()
        .delete(format!("{}/client/{id}", *GOTIFY))
        .header("X-Gotify-Key", token);

    let resp = client.send().await?;

    Ok(resp)
}